### ISSUE TITLE
Add support for the offline and online evalutation for SAM2

### ISSUE THREAD

**Issue by AlfredQin** [heyoeyo/muggled_sam#27]
Thanks for your great work. 
I tried to implement the evaluation code described in SAM2 paper but failed, the evaluation definition in SAM2 paper is shown below:

_Offline evaluation involves multiple passes over the entire video. We start with click prompts on the first
frame, segment the object throughout the entire video, and then in the next pass, we select the frame with the
lowest segmentation IoU w.r.t. the ground-truth as the new frame for prompting. The model then segments
the object again throughout the video based on all prompts received previously, until reaching a maximum of
Nframe passes (with one new prompted frame in each pass).
Online evaluation involves only one pass over the entire video. We start with click prompts on the first
frame and propagate the prompts across the video, pausing propagation when encountering a frame with a
low-quality prediction (IoU < 0.75 with ground-truth). We then add additional click prompts on the paused
frame to correct the segment on this frame and resume the propagation forward until reaching another low
quality frame with IoU < 0.75. This is repeated while the number of prompted frames is less than the
maximum Nframe . Unlike the previous offline evaluation, in this setting, the new prompts only affect the
frames after the current paused frame but not the frames before it._

Could you please help add support for this task ?

Thanks

**Comment by heyoeyo** [comment]
Hi, thanks for checking out the repo!

For the second case (online evaluation), I would guess something like this could work, at least as a starting point:

```python

from collections import deque
import torch
from muggled_sam.make_sam import make_sam_from_state_dict

# Assuming we have...
frames_list = []  # List of loaded images
gt_masks_list = []  # List of loaded ground-truth masks
calculate_mask_iou = lambda mask_a, mask_b: ...
get_corrective_prompts = lambda gt_mask, pred_mask: ...

# Set up model
_, model = make_sam_from_state_dict("/path/to/model.pth")
model.to(device="cuda", dtype=torch.bfloat16)

# Set up memory bank
prompt_mems = deque([])
prompt_ptrs = deque([])
prev_mems = deque([], maxlen=6)
prev_ptrs = deque([], maxlen=15)

# Process frame sequence
for frame, gt_mask in zip(frames_list, gt_masks_list):

    # Run video tracking
    encoded_imgs, _, _ = model.encode_image(frame)
    obj_score, best_mask_idx, mask_preds, mem_enc, obj_ptr = model.step_video_masking(
        encoded_imgs, prompt_mems, prompt_ptrs, prev_mems, prev_ptrs
    )

    # Encode more prompts when reaching low-iou frames
    best_mask_pred = mask_preds[:, best_mask_idx, :, :]
    if calculate_mask_iou(gt_mask, best_mask_pred) < 0.75:

        # Create new point prompts
        new_fg_points_list = get_corrective_prompts(gt_mask, best_mask_pred)

        # Encode & store new prompt memory
        new_mask, new_mem, new_ptr = model.initialize_video_masking(
            encoded_imgs,
            box_tlbr_norm_list=[],
            fg_xy_norm_list=new_fg_points_list,
            bg_xy_norm_list=[],
            mask_index_select=None,  # <- Use None to auto-select best mask
        )
        prompt_mems.appendleft(new_mem)
        prompt_ptrs.appendleft(new_ptr)

        # Report memory count
        print("Nframe:", len(prompt_mems))

    else:
        # Store frame memory
        # - Only on good IoU frames? And/or good obj_score?
        # - Not clear how this is handled...
        prev_mems.appendleft(mem_enc)
        prev_ptrs.appendleft(obj_ptr)

    pass
```

Though there are a few unknowns here:
- One is how the input images/masks are formatted, I've just written them in as lists (`frames_list` & `gt_masks_list`), but those would have to be loaded and indexed so the masks and frames are in the correct order
- There's also the iou calculation (i.e. `calculate_mask_iou` above), I'm not sure if that's bounding-box iou or (more likely) per-pixel iou with the masks scaled to some standard resolution?
- There's the `get_corrective_prompts` 'function'. I'd assume that's automated, but not sure how. It could be done manually/interactively, but then it seems the results wouldn't be repeatable...
- I'm not sure what happens when `Nframe` exceeds the maximum value (completely stop processing or just stop adding new prompts?) so it's just being printed out in the code above
- I'm not sure how the 'previous frame' memory encodings are handled (are they stored for every frame, or just for 'good iou' frames or do they use the `obj_score` to decide etc.)
- This assumes the 'best mask' is chosen based on the model's own predictions, though maybe it makes more sense to pick based on which one has the highest IoU with the ground-truth mask?
- This code doesn't include a separate first frame prompt, but without initial prompts, the iou is likely to be poor on the first frame so that should force a prompt anyways. But maybe the initial prompt is also loaded as part of the evaluation?
- Also, there's no output to any of this. I'd guess maybe the mask predictions should be stored for something or maybe just the number of frames processed is measured, again not sure

That's all just for the online evaluation. The offline version should be very similar, except with an extra loop around everything to handle multiple passes. Then then 'corrective prompt' part would come after processing the entire frame sequence (instead of being inside the frame sequence). After each loop, I assume they'd want to clear the 'previous frame' memory (e.g. calling `prev_mems.clear()` and `prev_ptrs.clear()`), but it sounds like they keep the prompt memory.
